### PR TITLE
SA Backend services - Tasks/10258

### PR DESCRIPTION
**Pull request by FuriousLlama**
# Fixes or Implements [AB#9710](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/9710) & [AB#10258](https://qslvic.visualstudio.com/304a1f8c-dace-4f85-adf3-bf563d5b3a39/_workitems/edit/10258)

* [x] Enhancement
* [ ] Bug
* [ ] Documentation

## Description

Added backend services to connect with Salesforce.

If you are reviewing this PR, please adhere to the guidelines as [documented](https://github.com/bcgov/healthgateway/wiki/prguidance).

## Testing

* [x] Unit Tests Updated
* [ ] Functional Tests Updated
* [ ] Not Required

### CODE REVIEW

**Review by tiagograf-AOT — APPROVED** [review]

File: Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs [hunk 1]
```diff
@@ -31,16 +32,18 @@ namespace Healthgateway.JobScheduler.Jobs
     /// <inheritdoc />
     public class NotificationSettingsJob : INotificationSettingsJob
     {
-        private const int ConcurrencyTimeout = 5 * 60; // 5 minutes
+        private const int concurrencyTimeout = 5 * 60; // 5 minutes
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs:35]: This change introduces a smell check.

File: Apps/JobScheduler/src/Jobs/CloseAccountJob.cs [hunk 2]
```diff
@@ -55,6 +57,9 @@ public class CloseAccountJob
         private readonly int hoursBeforeDeletion;
         private readonly string emailTemplate;
         private readonly string host;
+        private const string authConfigSectionName = "ClientAuthentication";
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CloseAccountJob.cs:60]: This doesn't follow constant variable naming standards

File: Apps/Medication/src/Delegates/SalesforeceDelegate.cs [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    using Microsoft.Extensions.Configuration;
+    using Microsoft.Extensions.Logging;
+
+    /// <summary>
+    /// Salesforce Implementation that retrieves Medication Requests.
+    /// </summary>
+    public class SalesforeceDelegate : IMedicationRequestDelegate
+    {
+        private const string AuthConfigSectionName = "ClientAuthentication";
+        private const string SalesForceConfigSectionKey = "Salesforce";
```
Comment (MikeLyttle) [inline Apps/Medication/src/Delegates/SalesforeceDelegate.cs:42]: Suggest renaming constant to `SalesforceConfigSectionKey`.

File: Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs [hunk 4]
```diff
@@ -31,16 +32,18 @@ namespace Healthgateway.JobScheduler.Jobs
...
         private const string JobConfigKey = "NotificationSettings";
         private const string JobEnabledKey = "Enabled";
-
+        private const string authConfigSectionName = "ClientAuthentication";
```
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs:38]: Another smell

File: Apps/Common/src/AccessManagement/Authorization/Policy/MedicationPolicy.cs [hunk 5]
```diff
@@ -21,13 +21,23 @@ namespace HealthGateway.Common.AccessManagement.Authorization.Policy
     public static class MedicationPolicy
     {
         /// <summary>
-        /// Policy which allows the reading of the identified patient.
+        /// Policy which allows the reading medication statements.
```
Comment (MikeLyttle) [inline Apps/Common/src/AccessManagement/Authorization/Policy/MedicationPolicy.cs:24]: /// Policy which allows the reading _of_ medication statements.

File: Apps/JobScheduler/src/Jobs/CloseAccountJob.cs (hunk 2 above)
Comment (sslaws) [inline Apps/JobScheduler/src/Jobs/CloseAccountJob.cs:60]: this looks like a duplicate of the variable declared on 44

**Review by sslaws — CHANGES_REQUESTED** [review]

File: Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+                Username = "TEST_USERNAME",
+            };
+            Dictionary<string, string> configurationParams = new Dictionary<string, string>
+            {
+                { "Salesforce:Endpoint", endpoint },
+                { "Salesforce:TokenUri", tokenUri.ToString() },
+                { "Salesforce:ClientAuthentication:ClientId", tokenRequest.ClientId },
+                { "Salesforce:ClientAuthentication:ClientSecret", tokenRequest.ClientSecret },
+                { "Salesforce:ClientAuthentication:Username", tokenRequest.Username },
+                {"Salesforce:ClientAuthentication:Password", tokenRequest.Password },
```
Comment (MikeLyttle) [inline Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs:68]: Formatting - missing space.

File: Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+                mockHttpClient.Object,
+                configuration,
+                mockAuthenticationDelegate.Object);
+
+
+            // Test
+            RequestResult<IList<MedicationRequest>> response = Task.Run(async () =>
+                                                                    await medDelegate.GetMedicationRequestsAsync(phn)).Result;
+
+
```
Comment (MikeLyttle) [inline Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs:103]: Formatting: extra whitespace.

File: Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs (hunk 6 above)
Comment (MikeLyttle) [reply Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs:68]: Also applies to similar sections in this file.

File: Apps/JobScheduler/src/Jobs/CloseAccountJob.cs (hunk 2 above)
Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/CloseAccountJob.cs:60]: good catch. Removed

File: Apps/Common/src/AccessManagement/Authorization/Policy/MedicationPolicy.cs (hunk 5 above)
Comment (FuriousLlama) [reply Apps/Common/src/AccessManagement/Authorization/Policy/MedicationPolicy.cs:24]: fixed

File: Apps/Medication/src/Delegates/SalesforeceDelegate.cs (hunk 3 above)
Comment (FuriousLlama) [reply Apps/Medication/src/Delegates/SalesforeceDelegate.cs:42]: sounds good. Fixed

File: Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs (hunk 6 above)
Comment (FuriousLlama) [reply Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs:68]: fixed

File: Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs (hunk 7 above)
Comment (FuriousLlama) [reply Apps/Medication/test/unit/Delegates/SalesforceDelegate_Test.cs:103]: fixed

File: Apps/JobScheduler/src/Jobs/CloseAccountJob.cs (hunk 2 above)
Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/CloseAccountJob.cs:60]: it does now!

File: Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs (hunk 1 above)
Comment (FuriousLlama) [reply Apps/JobScheduler/src/Jobs/NotificationSettingsJob.cs:35]: fixed

**Review by sslaws — APPROVED** [review]

**Review by MikeLyttle — APPROVED** [review]

**Pull request merged**